[KEP] Support new ProvisioningRequest's conditions (#2042)
kubernetes-sigs/kueue: keps/1136-provisioning-request-support/README.md

> [Two phase admission process](https://github.com/kubernetes-sigs/kueue/tree/main/keps/993-two-phase-admission)

Reviewer: this paragraph is no longer true, you can remove it.

Author: Done

To understand what details should it put into `ProvisioningRequest` the controller
will also need to watch `ProvisioningRequestConfigs`.

* Watch all changes CA makes to `ProvisioningRequests`. If the `ProvisioningRequest's` conditions are set to:
  - `Provisioned=false` controller should surface information about ProvisioningRequest's ETA. It should emit an event regarding that and for every ETA change.
  - `Provisioned=true` controller should mark the AdmissionCheck as `Ready` and propagate the information about `ProvisioningRequest` name to
workload pods - [KEP #1145](https://github.com/kubernetes-sigs/kueue/blob/main/keps/1145-additional-labels/kep.yaml) under `"cluster-autoscaler.kubernetes.io/consume-provisioning-request"`.
  - `Failed=true` controller should retry AdmissionCheck with respect to the `RetryConfig` configuration, or mark the AdmissionCheck as `Rejected`

>   - `BookingExpired=true` controller should check if `ProvisioningRequest` needs recreating, and retry if needed with respect to the `RetryConfig` configuration.

Reviewer 2: nit: Maybe worth specify when ProvisioningRequest needs to be recreated?

Author: Done, thanks

>   - `BookingExpired=true` controller should check if `ProvisioningRequest` needs recreating - when the provisioned nodes have no pods scheduled, and retry if needed with respect to the `RetryConfig` configuration.

Reviewer 2: I don't think we need to check provisioned nodes. For example for Check-capacity ProvisioningClass there are no provisioned nodes. 
I think ProvisioningRequest should be recreated if the pods weren't created yet or if the pods were created but not all of them were scheduled during some time.

Reviewer: +1 on an even simpler approach of checking if the job is unsuspended.

Maybe start a section in Risks and Mitigations to describe when the race condition that could happen. IIRC, it could only happen if the user has more than one admission check or when creating a massive job, where the job controller is not able to create pods fast enough
For the first case, we have WaitForPodsReady as a possible mitigation (after the timeout, the job is evicted and retried).
For the second case, the cluster administrator should probably size their control plane appropriately (bigger VMs and/or bigger qps for the controller creating pods).

To understand what details should it put into `ProvisioningRequest` the controller
will also need to watch `ProvisioningRequestConfigs`.

* Watch all changes CA makes to `ProvisioningRequests`. If the `ProvisioningRequest's` conditions are set to:
  - `Provisioned=false` controller should surface information about ProvisioningRequest's ETA. It should emit an event regarding that and for every ETA change.
  - `Provisioned=true` controller should mark the AdmissionCheck as `Ready` and propagate the information about `ProvisioningRequest` name to
workload pods - [KEP #1145](https://github.com/kubernetes-sigs/kueue/blob/main/keps/1145-additional-labels/kep.yaml) under `"cluster-autoscaler.kubernetes.io/consume-provisioning-request"`.
  - `Failed=true` controller should retry AdmissionCheck with respect to the `RetryConfig` configuration, or mark the AdmissionCheck as `Rejected`

>   - `BookingExpired=true` if a Workload is not `Admitted` controller should act the same as for `Failed=true`.

Reviewer: ```suggestion
  - `BookingExpired=true` if a Workload is not `Admitted`, the controller should act the same as for `Failed=true`.
```

>   - `CapacityRevoked=true` if a Workload is not `Finished` controller should mark it as `Inactive`, which will evict it.

Reviewer: ```suggestion
  - `CapacityRevoked=true` if a Workload is not `Finished`, the controller should mark it as `Inactive`, which will evict it.
```

Reviewer: Maybe something for the implementation, not the KEP, but we should somehow communicate why the Workload was deactivated.
Can we leave a message in the admission checks slice or something?

>     Additionally, an event should be emitted to signalize this happening. This can happen only if a user uses `batch.v1/Job` and

Reviewer: ```suggestion
    Additionally, an event should be emitted to signalize this happening. This can happen only if the job
    allows for retries, for example, in the case of `batch.v1/Job`, when the user
```
